### PR TITLE
Allow extensions to access all terminals

### LINKED ISSUE #13267: Allow extensions to access all terminals via API
To give extensions more control over the terminal panel I'm proposing the following:

``` ts
export namespace window {
  export const onDidOpenTerminal: Event<Terminal>;
  export let terminals: Terminal[];
}
```

A solid use case for this is to represent the terminals as tabs somewhere in the UI https://github.com/Tyriar/vscode-terminal-tabs

### PR DESCRIPTION
**Pull request by Tyriar**
New proposed APIs:

```ts
export namespace window {
	/**
	 * The currently active terminals or an empty array.
	 *
	 * @readonly
	 */
	export let terminals: Terminal[];

	/**
	 * An [event](#Event) which fires when a terminal has been created, either through the
	 * [createTerminal](#window.createTerminal) API or commands.
	 */
	export const onDidOpenTerminal: Event<Terminal>;
}
```

Changes to existing behavior:

- `Terminal.name` and `Terminal.processId` can be accessed on a disposed terminal

See usage example at https://github.com/Microsoft/vscode-extension-samples/blob/master/terminal-sample/src/extension.ts

Fixes #13267
Part of #46192

/cc @dbaeumer @weinand @IlyaBiryukov @alpaix

### CODE REVIEW

File: src/vs/vscode.proposed.d.ts [hunk 1]
```diff
@@ -769,4 +769,23 @@ declare module 'vscode' {
...
+		 *
+		 * @readonly
+		 */
+		export let terminals: Terminal[];
+
+		/**
+		 * An [event](#Event) which fires when a terminal has been created, either through the
+		 * [createTerminal](#window.createTerminal) API or commands.
+		 */
+		export const onDidOpenTerminal: Event<Terminal>;
```
Comment (jrieken) [inline src/vs/vscode.proposed.d.ts]: Think about a `TerminalOpenEvent` so that you a chance of adding more stuff in the future..

Comment (jrieken) [reply src/vs/vscode.proposed.d.ts]: Then, should open close with close?

Comment (Tyriar) [reply src/vs/vscode.proposed.d.ts]: `onDidCloseTerminal` already exists as an `Event<Terminal>`:

https://github.com/Microsoft/vscode/blob/da9720b39b1a912ce8009587731d824a890797ad/src/vs/vscode.d.ts#L4905-L4908

Comment (jrieken) [reply src/vs/vscode.proposed.d.ts]: missed that 🤦‍♂️

Comment (Tyriar) [reply src/vs/vscode.proposed.d.ts]: Different file :wink:

**Pull request merged**